compiler/codegen/CodeGenPrep.cpp

```diff
@@ -177,6 +179,125 @@ OMR::CodeGenerator::lowerTreesPreChildrenVisit(TR::Node * parent, TR::TreeTop *
+ * other threads.
+ *
+ * @param comp pointer to the compilation object
+ * @param node the node being lowered
+ * @param tt the TreeTop anchoring the node
+ */
+static void
+lowerNewValue(TR::Compilation *comp, TR::Node *node, TR::TreeTop *tt)
+   {
+   // Transmute newvalue node into new.
+   // Importantly, the helper symref of the newvalue node is preserved.
+   TR::Node::recreate(node, TR::New);
+
+   auto* valueClass = static_cast<TR_OpaqueClassBlock *>(node->getFirstChild()->getSymbol()->getStaticSymbol()->getStaticAddress());
+   const TR::TypeLayout* typeLayout = comp->typeLayout(valueClass);
+   TR::IL il;
+
+   // cursors to iterate over the treetops that compute field values and store those values into fields
+   auto* fieldValueTreeTopCursor = tt->getPrevTreeTop();
+   auto* fieldStoreTreeTopCursor = tt;
+
+   // cache the treetop that must go at the end of the new sequence
+   auto* nextTreeTop = tt->getNextTreeTop();
+
+   // Iterate over very child that sets a field, anchoring computations
+   // before the allocation node and anchoring stores to fields after.
+   //
+   // The first child is skipped as it's only used to specify the type
+   // of the value being constructed. The second child is the one that
+   // actually initializes the first field.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      TR::Node* fieldValueNode = node->getChild(i);
+      node->setChild(i, NULL);
+
+      // anchor calculation of the field's value
+      auto* ttNode = TR::Node::create(TR::treetop, 1);
+      ttNode->setFirst(fieldValueNode);
+      fieldValueTreeTopCursor->join(TR::TreeTop::create(comp, ttNode));
+      fieldValueTreeTopCursor = fieldValueTreeTopCursor->getNextTreeTop();
+
+      // generate store to the field
+      const TR::TypeLayoutEntry& fieldEntry = typeLayout->entry(i - 1);
+      auto* symref = comp->getSymRefTab()->findOrFabricateShadowSymbol(valueClass,
+                                                                       fieldEntry._datatype,
+                                                                       fieldEntry._offset,
+                                                                       fieldEntry._isVolatile,
+                                                                       fieldEntry._isPrivate,
+                                                                       fieldEntry._isFinal,
+                                                                       fieldEntry._fieldname,
+                                                                       fieldEntry._typeSignature
+                                                                       );
+      const auto storeOpCode = il.opCodeForIndirectStore(fieldValueNode->getDataType());
+      auto* storeNode = TR::Node::createWithSymRef(storeOpCode, 2, symref);
+      storeNode->setAndIncChild(0, node);
+      storeNode->setAndIncChild(1, fieldValueNode);
```

Comment:
Yes, the increment is needed here because the node is also set as a child of the `ttNode` above. The ref count is correct at that earlier point so no increment is needed there, but it is needed at this later point.

Reply:
Oops, looks like you're right. No idea how I managed to miss that this node was being used twice.